pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -607,16 +607,18 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun, get
 				return controller.NewPermanentError(err)
 			}
 		}
+	}
 
-		if !c.isAffinityAssistantDisabled(ctx) {
-			// create Affinity Assistant (StatefulSet) so that taskRun pods that share workspace PVC achieve Node Affinity
-			if err = c.createAffinityAssistants(ctx, pr.Spec.Workspaces, pr, pr.Namespace); err != nil {
-				logger.Errorf("Failed to create affinity assistant StatefulSet for PipelineRun %s: %v", pr.Name, err)
-				pr.Status.MarkFailed(ReasonCouldntCreateAffinityAssistantStatefulSet,
-					"Failed to create StatefulSet for PipelineRun %s/%s correctly: %s",
-					pr.Namespace, pr.Name, err)
-				return controller.NewPermanentError(err)
-			}
+	// Make an attempt to create Affinity Assistant if it does not exist
+	// if the Affinity Assistant already exists, handle the possibility of assigned node going down by deleting the pod
+	if !c.isAffinityAssistantDisabled(ctx) {
```

Comment:
> It seems like this would only handle unschedulable nodes at the time when the affinity assistant is being created (before the first taskrun),

I have moved out the call to `createAffinityAssistants` outside of the [check](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L583
) `IsBeforeFirstTaskRun` such that an attempt is made on every reconciler cycle. 

Yeah, I tested these changes with the deadlock scenarios you have captured in https://github.com/tektoncd/pipeline/issues/6586 and these changes addresses the `pending` state of the pod creation for the  `good-morning-run-afkjf-last-pod`. These changes creates an affinity assistant pod on a healthy node (schedulable) along with the `last-pod`.